src/snowflake/snowpark/_internal/udf_utils.py

```diff
@@ -186,95 +186,120 @@ def extract_return_type_from_udtf_type_hints(
                 )
             python_types_dict = get_python_types_dict_for_udtf(
-                process_types_dict or {}, end_partition_types_dict or {}
-            )
-        elif object_type in (TempObjectType.FUNCTION, TempObjectType.PROCEDURE):
-            python_types_dict = retrieve_func_type_hints_from_source(
-                filename, func_name
+                process_types_dict, end_partition_types_dict
             )
         else:
+            python_types_dict = get_type_hints(func)
+    except TypeError:
+        # if we fail to run get_type_hints on a function (a TypeError will be raised),
+        # return empty type dict. This will fail for functions like numpy.ufunc
+        # (e.g., get_type_hints(np.exp))
+        pass
+    return python_types_dict or {}
+
+
+def _extract_types_dict_from_file(
+    filename: str,
+    func_name: str,
+    object_type: TempObjectType,
+) -> Dict[str, Any]:
+    python_types_dict: Optional[Dict[str, Any]] = None
+    if not is_local_python_file(filename):
+        pass
+    elif object_type == TempObjectType.AGGREGATE_FUNCTION:
+        accumulate_hints = (
+            retrieve_func_type_hints_from_source(
+                filename, AGGREGATE_FUNCTION_ACCULUMATE_METHOD, class_name=func_name
+            )
+            or {}
+        )
+        finish_hints = (
+            retrieve_func_type_hints_from_source(
+                filename, AGGREGATE_FUNCTION_FINISH_METHOD, class_name=func_name
+            )
+            or {}
+        )
+        python_types_dict = get_python_types_dict_for_udaf(
+            accumulate_hints, finish_hints
+        )
+    elif object_type == TempObjectType.TABLE_FUNCTION:
+        process_types_dict = retrieve_func_type_hints_from_source(
+            filename, TABLE_FUNCTION_PROCESS_METHOD, class_name=func_name
+        )
+        end_partition_types_dict = retrieve_func_type_hints_from_source(
+            filename, TABLE_FUNCTION_END_PARTITION_METHOD, class_name=func_name
+        )
+        if process_types_dict is None and end_partition_types_dict is None:
             raise ValueError(
-                f"Expecting FUNCTION, PROCEDURE, TABLE_FUNCTION, or AGGREGATE_FUNCTION as object_type, got {object_type}"
+                f"Neither {func_name}.{TABLE_FUNCTION_PROCESS_METHOD} or {func_name}.{TABLE_FUNCTION_END_PARTITION_METHOD} could be found from {filename}"
             )
+        python_types_dict = get_python_types_dict_for_udtf(
+            process_types_dict or {}, end_partition_types_dict or {}
+        )
+    elif object_type in (TempObjectType.FUNCTION, TempObjectType.PROCEDURE):
+        python_types_dict = retrieve_func_type_hints_from_source(filename, func_name)
+    else:
+        raise ValueError(
+            f"Expecting FUNCTION, PROCEDURE, TABLE_FUNCTION, or AGGREGATE_FUNCTION as object_type, got {object_type}"
+        )
+    return python_types_dict or {}
+
+
+def get_types_from_type_hints(
```

Comment:
This function was overly verbose and reused variable names even though they had slightly different meanings. Separating out the differing logic makes it a little more readable and clears up some mypy errors.

Reply:
This is great, thanks for refactoring this!